improve repo setup and replicability

crispy-wonton opened the pull request:
Fixes #2 

Changes:

Overall changes to make the script more streamlined and automated.

- Update README with clearer and updated instructions and with files used in historical analysis
- Add arg parser so user can input local_data_dir, EPC batch, and MCS batch in args instead of changing script. Script will load MCS data directly from S3 and will automatically download specified EPC batch if not found in local directory.
- Fix issues with altair by saving as .html rather than .png
- Update mcs_epc getter functions to work with newer batches that do not have "end_user_installation_type" column. Functions will still work with older batches that do have that column.
- Update requirements.txt

**NOTE: the code requires new functions added in [asf_core_data PR 102](https://github.com/nestauk/asf_core_data/pull/102) to run**

During review, please could you:
- Give your thoughts on the use of config file for EPC processing version
- Indicate if you feel additional arguments would be useful
- Follow the instructions of the README to check it's clear and test the code


# Checklist:

- [y ] I have refactored my code out from `notebooks/`
- [y ] I have checked the code runs
- [y ] I have tested the code
- [y ] I have run `pre-commit` and addressed any issues not automatically fixed
- [y ] I have merged any new changes from `dev`
- [y ] I have documented the code
  - [y ] Major functions have docstrings
  - [y ] Appropriate information has been added to `README`s
- [y ] I have explained this PR above
- [y ] I have requested a code review

crispy-wonton committed:
update README

improve and correct set up instructions and update output file extensions to html.

crispy-wonton committed:
update local_data_dir in getters

update local_data_dir to use relative file path from config base.yaml so config file can be updated by users rather than editing scripts.

crispy-wonton committed:
update requirements.txt with versions

crispy-wonton committed:
change outputs to .html files

change output graphs from produce_plots.py from .png to .html files to avoid altair CalledProcessError

crispy-wonton committed:
fix introduced typo in requirements.txt

crispy-wonton committed:
automate all data downloads/loading in script

- add argparser to allow user to specify local_data_dir, epc-, and mcs-batch
- add function to load MCS data directly from S3 bucket
- add function to check for EPC data locally and download from S3 if not located
- update existing functions to work with above
- add global parameters for EPC processing version to base.yaml config file

crispy-wonton committed:
update README

update instructions in README to explain new way of running script and to record batches for historical analyses

crispy-wonton committed:
update requirements.txt with argpase

sofiapinto commented on README.md:
```diff
@@ -9,30 +9,31 @@ The remainder of the charts in the response can be produced from code in the rep
   - Setup the conda environment
   - Configure `pre-commit`
-- Run `direnv allow`
+  - Install packages listed in `requirements.txt`
```
This shouldn't need to happen actually. When you do make install, `pip install -r requirements.txt` should be ran automatically. I'm not sure that's not working for this repo. The pip install solves it, but it's weird.

Because this is not working well might be worth keeping the `pip install -r requirements.txt` so people know how to do it.

sofiapinto replied:
okay, all working with `make install` so no need to pip install the requirements

sofiapinto commented on README.md:
```diff
@@ -9,30 +9,31 @@ The remainder of the charts in the response can be produced from code in the rep
+  default to downloading/loading the newest data from S3, respectively. Run `python asf_welsh_energy_consultation/analysis/produce_plots.py -h` for more info.
 
-  - Follow the instructions (here)[https://github.com/altair-viz/altair_saver/#nodejs] - you may just need to run `conda install -c conda-forge vega-cli vega-lite-cli`
+The script should generate the following six plots:
```
Maybe add something about where they are going to be stored

sofiapinto commented on README.md:
````diff
@@ -53,6 +54,18 @@ outputs/
 ├─ figures/ - where charts are saved
 ```
 
+## Historical analysis
````
Great! This is going to be super handy. Don't forget to add the details of the current analysis (in this or in a new PR, as you see fit).

crispy-wonton replied:
yep will add in a separate PR

sofiapinto commented on asf_welsh_energy_consultation/analysis/produce_plots.py:
evil altair 😈

sofiapinto commented on asf_welsh_energy_consultation/config/base.yaml:
```diff
@@ -0,0 +1,3 @@
+epc_data_config:
+  epc_processing_version: "preprocessed"
```
would there ever be a case where we would need a different version that's not the `preprocessed`?

crispy-wonton replied:
I dont think so, so this is a bit redundant. However I do think it makes it slightly easier to identify which version is being used. asf_senedd_response uses a different version so may not be useful when repos are merged.

sofiapinto replied:
happy for you to leave it as you prefer!

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -4,22 +4,111 @@
 
-LOCAL_DATA_DIR = "/Users/chris.williamson/Documents/ASF_data"
+
+def create_argparser():
```
great 🥳

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -62,12 +151,13 @@ def get_mcs_domestic():
-        mcs["end_user_installation_type"]
-    )
-    mcs = mcs.drop(columns=["end_user_installation_type"])
+    if "end_user_installation_type" in mcs.columns:
+        mcs["installation_type"] = mcs["installation_type"].fillna(
+            mcs["end_user_installation_type"]
+        )
+        mcs = mcs.drop(columns=["end_user_installation_type"])
```
This is a good fix, thanks for that. Maybe you could just add some information explaining why we're doing this (as in a small comment). But basically:
- we always receive two variables "end_user_installation_type"  and "installation_type":  they correspond to the same variable, but one of them is null up to some date and the other is null after a specific date; initially, we didn't do anything to these variables but [a couple of months a go we integrated this `fillna` step in the pipeline](https://github.com/nestauk/asf_core_data/issues/77) - so older batches will require this processing step here, and the newest batches won't so this works perfectly.

crispy-wonton replied:
added

sofiapinto commented on requirements.txt:
This is great! The only thing I had to change was the asf_core_data branch, for it to work with the functions you defined (since they haven't been merged to `dev` yet), i.e. using:

`asf_core_data@ git+ssh://git@github.com/nestauk/asf_core_data.git@100_fix_getting_processed_mcs_installations_data_bug`

Obviously this change won't be needed once you merge the other branch to dev in `asf_core_data`. Just give it a try to see if all still runs fine.

crispy-wonton replied:
branch merged now so should work ! :D

sofiapinto commented on README.md:
```diff
@@ -9,30 +9,31 @@ The remainder of the charts in the response can be produced from code in the rep
+- Navigate to your local repo folder
 - Checkout the correct branch if not working on dev
-- Run `make install` to configure the development environment:
+- Run `direnv allow`
+- Run `make install` to configure the development environment. This will:
```
isn't it the other way around?

crispy-wonton replied:
not sure. will test with someone installing for first time and then update as necessary

sofiapinto commented on README.md:
```diff
@@ -9,30 +9,31 @@ The remainder of the charts in the response can be produced from code in the rep
+- Run `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>`. You need to specify the path to the local
+  directory where your local copy of the EPC data is/will be saved by replacing `<YOUR_LOCAL_DIR>` with the path. You can specify which
+  batch of EPC data to download and MCS data to load from S3 by passing the `--epc_batch` and `--mcs_batch` arguments, both
+  default to downloading/loading the newest data from S3, respectively. Run `python asf_welsh_energy_consultation/analysis/produce_plots.py -h` for more info.
```
nice 🤓

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -139,14 +229,55 @@ def get_rurality():
+        epc_batch.upper().replace("COMPLETE", "complete"),
+    )
+
+    local_epc_path = os.path.join(
+        local_epc_batch_path, f"EPC_GB_{epc_processing_version}.csv"
+    )
```
When I print this path it gives me:
`/Users/anasofiapinto/Documents/ASF data/outputs/EPC/preprocessed_data/NEWEST/EPC_GB_preprocessed.csv`

hence, it's looking for the file in an incorrect path, that's why it can't find it. It's trying to find it in 'NEWEST' instead of the correct quarter folder.

It should be trying to find this path:
`/Users/anasofiapinto/Documents/ASF data/outputs/EPC/preprocessed_data/2023_Q2_complete/EPC_GB_preprocessed.csv`

sofiapinto replied:
not sure if the same happens when you specify the batch - haven't tested it

crispy-wonton replied:
have updated this and tested locally. should work now but please do run again to test :)

sofiapinto replied:
it worked now :)

sofiapinto commented on README.md:
```diff
@@ -9,30 +9,31 @@ The remainder of the charts in the response can be produced from code in the rep
 
-  - `make inputs-pull` will pull the zipped data from S3 and put it in `/inputs`
-  - Download `mcs_installations_230315.csv` and `mcs_installations_epc_full_230315.csv` from `asf-core-data` S3 bucket (they can be found under `outputs/MCS/`) and add them to `inputs/data/`- Perform additional setup in order to save plots:
+- Run `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>`. You need to specify the path to the local
+  directory where your local copy of the EPC data is/will be saved by replacing `<YOUR_LOCAL_DIR>` with the path. You can specify which
```
Wondering if you should say something more about the local directory something like (or similar):

_If you haven't created a local directory for ASF core data, you can create a folder in your Documents called "ASF data". Your local directory will then be something like `/Users/your_username/Documents/ASF data/`_

crispy-wonton replied:
good idea

sofiapinto reviewed (changes requested):
Hey @crispy-wonton,

Thanks a lot for this PR! This is a great improvement to the repo 🥳

I had a bit of initial trouble with the set up. I always do with this repo... not sure why! Maybe it's a problem with the way the cookiecutter was setup initially? After you merge this branch to `dev`, might be worth asking someone else (maybe Aidan) who hasn't installed the repo previously to run through the steps in the `README.md` and make sure they don't run into any problem - so that someone does it from scratch.

But anyway, this is what I did (besides looking through the changes in the scripts):

- I ran through the steps in the `README.md` - it's weird because `make install` should install the packages but that's not happening so I had to run `pip install - r requirements.txt`. 
- Then I noticed that the `asf_core_data` package was being installed with the dev branch, so I changed that to enable the testing of the new functions (I left a comment about that)|;
- When trying to run the `produce_plots.py` script I got an error ([this has happened to other people in the past](https://data-science-nesta.slack.com/archives/CAYNV0XFZ/p1669810366657559)) saying" "No module named asf_welsh_energy_consultation". This was fixed by running `export PYTHONPATH=$PWD` in the command line;
- I was finally able to run `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir <YOUR_LOCAL_DIR>` with my local dir without trouble.

Weirdly, when I ran the script, it said that it couldn't find the newest batch in my local folder.. but it's there, so there's a small mistake there that needs fixing.


Btw I also get this warning:
```
/Users/anasofiapinto/opt/anaconda3/envs/asf_welsh_energy_consultation/lib/python3.8/site-packages/fsspec/registry.py:271: UserWarning: Your installed version of s3fs is very old and known to cause
severe performance issues, see also https://github.com/dask/dask/issues/10276

To fix, you should specify a lower version bound on s3fs, or
update the current installation.
```


Happy to re-run things again after the fix :)

crispy-wonton committed:
fix check_local_epc function in getters

check_local_epc function updated so that it functions properly if batch specified is 'newest' (or variation).
documentation also added to get_mcs_domestic function

crispy-wonton committed:
update README

provide further clarity as requested in review

crispy-wonton committed:
update requirements to fix s3fs warning

crispy-wonton committed:
remove additional altair installation instructions

remove additional altair installation instructions because they are only required for saving to .png files which is not used so no longer required

sofiapinto reviewed (commented):
hey @crispy-wonton ,

- I've ran `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir "/Users/anasofiapinto/Documents/ASF data/"` and it identified the correct path to the local data now 🥳
-  Also tested `python asf_welsh_energy_consultation/analysis/produce_plots.py --local_data_dir "/Users/anasofiapinto/Documents/ASF data/" --epc_batch "2022_Q3_complete" --mcs_batch "230315"` - not that these specific `epc_batch` and `mcs_batch` make sense together, but I just wanted to test if choosing specific batches worked well :)

Both ran fine!

Also now I had no problem with `make install`, it installed the packages properly.

Something I noticed when looking at the plots this time: some don't use the Nesta blue.

crispy-wonton committed:
update documentation for get_electric_tenure function for clarity

crispy-wonton committed:
replace print with logger
